File: src/components/swipe.rs
//! Swipeable list rows.
use super::Color;
use super::gesture::{Gesture, GestureScript, gesture_command, use_gesture};
use super::list::{InList, InSwipeRow};
use crate::state::use_element_id;
use crate::theme::{classes, merge_classes, use_strings};
use dioxus::prelude::*;

/// Which edge of a row a swipe uncovers.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum SwipeSide {
    /// The leading edge, uncovered by dragging toward the trailing side.
    Start,
    /// The trailing edge, the usual home of destructive actions.
    End,
}

/// What swiping toward one edge does once it passes its threshold.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Default)]
pub enum SwipeBehavior {
    /// Uncover the actions and hold them open until one is pressed or the
    /// row is swiped closed.
    #[default]
    Reveal,
    /// Run a callback on release, as with swipe-to-archive. The first action
    /// stretches across the row as the swipe passes the threshold.
    Activate,
    /// Slide the row away and remove it.
    Dismiss,
}

impl SwipeBehavior {
    fn as_str(self) -> &'static str {
        match self {
            SwipeBehavior::Reveal => "reveal",
            SwipeBehavior::Activate => "activate",
            SwipeBehavior::Dismiss => "dismiss",
        }
    }
}

/// The live state of a swipe, passed to [`SwipeItem`]'s callbacks.
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct SwipeState {
    /// Which edge is being uncovered.
    pub side: SwipeSide,
    /// Horizontal displacement of the row, in pixels.
    pub offset: f64,
    /// `offset` as a fraction of the threshold distance; `1.0` is fully open.
    pub ratio: f64,
    /// Whether releasing now would activate or dismiss.
    pub committed: bool,
}

/// Runs every row's gesture inside the webview, so Rust hears only how each
/// one ended. See [`super::gesture`] for why.
const SWIPE_SCRIPT: GestureScript = GestureScript {
    name: "g3-ui.swipe",
    source: include_str!("swipe.js"),
};

/// What the script reports about one row.
#[derive(Clone, Copy, Debug, PartialEq)]
enum SwipeEvent {
    /// The row came to rest at this offset: closed, or open on one edge.
    Rest(f64),
    /// The row moved; only reported when someone listens.
    Move(SwipeState),
    Activate(SwipeState),
    Dismiss(SwipeState),
    LongPress,
}

impl SwipeEvent {
    /// Reads `[offset, ratio, committed]`.
    fn parse(gesture: &Gesture) -> Option<Self> {
        let (offset, ratio) = (gesture.value(0), gesture.value(1));
        let committed = gesture.value(2) != 0.0;
        let state = || {
            side_of(offset).map(|side| SwipeState {
                side,
                offset,
                ratio,
                committed,
            })
        };
        Some(match gesture.kind.as_str() {
            "rest" => SwipeEvent::Rest(offset),
            "move" => SwipeEvent::Move(state()?),
            "activate" => SwipeEvent::Activate(state()?),
            "dismiss" => SwipeEvent::Dismiss(state()?),
            "long-press" => SwipeEvent::LongPress,
            _ => return None,
        })
    }
}

fn side_of(offset: f64) -> Option<SwipeSide> {
    if offset > 0.0 {
        Some(SwipeSide::Start)
    } else if offset < 0.0 {
        Some(SwipeSide::End)
    } else {
        None
    }
}

/// A button uncovered by swiping a [`SwipeItem`].
#[component]
pub fn SwipeAction(
    /// Color. Defaults to [`Color::Neutral`].
    color: Option<Color>,
    /// Accessible name, for icon-only actions.
    aria_label: Option<String>,
    /// Called when pressed.
    onclick: Option<EventHandler<MouseEvent>>,
    /// Extra classes for the action.
    class: Option<String>,
    children: Element,
) -> Element {
    let color_cls = format!(
        "g3-swipe-action-{}",
        color.unwrap_or(Color::Neutral).as_str()
    );
    rsx! {
        button {
            class: merge_classes(classes(["g3-swipe-action", &color_cls]), class.as_deref()),
            r#type: "button",
            aria_label,
            onclick: move |event| {
                if let Some(onclick) = onclick {
                    onclick.call(event);
                }
            },
            {children}
        }
    }
}

/// A list row with actions behind it, uncovered by swiping. Like Ionic's
/// `ion-item-sliding`.
///
/// Put an [`Item`](crate::Item) inside and [`SwipeAction`]s in
/// `start_actions` or `end_actions`. Each edge has its own behaviour, so a
/// row can archive on a swipe one way and reveal buttons the other way.
/// Keyboard and screen reader users reach the actions through a "Show
/// actions" button that appears on focus, or with the arrow keys; Escape
/// closes them.
///
/// ```
/// # use dioxus::prelude::*;
/// # use g3_ui::prelude::*;
/// # fn demo() -> Element {
/// # let id = 7_u32;
/// # fn archive(_id: u32) {}
/// # fn pin(_id: u32) {}
/// # fn delete(_id: u32) {}
/// rsx! {
///     SwipeItem {
///         start_behavior: SwipeBehavior::Activate,
///         start_actions: rsx! { SwipeAction { color: Color::Success, "Archive" } },
///         on_activate: move |_| archive(id),
///         end_actions: rsx! {
///             SwipeAction { onclick: move |_| pin(id), "Pin" }
///             SwipeAction { color: Color::Danger, onclick: move |_| delete(id), "Delete" }
///         },
///         Item { label: "Round 12" }
///     }
/// }
/// # }
/// ```
#[component]
pub fn SwipeItem(
    /// Actions uncovered by swiping toward the trailing edge.
    start_actions: Option<Element>,
    /// Actions uncovered by swiping toward the leading edge.
    end_actions: Option<Element>,
    /// What swiping far toward the trailing edge does. Defaults to
    /// [`SwipeBehavior::Reveal`].
    start_behavior: Option<SwipeBehavior>,
    /// What swiping far toward the leading edge does. Defaults to
    /// [`SwipeBehavior::Reveal`].
    end_behavior: Option<SwipeBehavior>,
    /// Turn swiping off.
    disabled: Option<bool>,
    /// Let a mouse drag swipe too. Defaults to `true`; turn it off to keep
    /// the gesture for touch while desktop users use the actions button.
    mouse_swipe: Option<bool>,
    /// Called as the row moves.
    on_swipe: Option<EventHandler<SwipeState>>,
    /// Called when a [`SwipeBehavior::Activate`] swipe is released past its
    /// threshold. [`SwipeState::side`] says which edge.
    on_activate: Option<EventHandler<SwipeState>>,
    /// Called once a [`SwipeBehavior::Dismiss`] row has slid away and
    /// collapsed. Remove the row here.
    on_dismiss: Option<EventHandler<SwipeState>>,
    /// Called when the row is held without moving.
    on_long_press: Option<EventHandler<()>>,
    /// Extra classes for the row.
    class: Option<String>,
    children: Element,
) -> Element {
    let strings = use_strings();
    let row_id = use_element_id("swipe", None);
    let in_list = use_hook(|| try_consume_context::<InList>().is_some());
    use_context_provider(|| InSwipeRow);
    let start_behavior = start_behavior.unwrap_or_default();
    let end_behavior = end_behavior.unwrap_or_default();
    let has_start = start_actions.is_some();
    let has_end = end_actions.is_some();
    let disabled = disabled.unwrap_or(false);
    let mouse_swipe = mouse_swipe.unwrap_or(true);
    // Where the row rests, as the script last reported it. Only used to keep
    // closed actions out of reach; the live offset never leaves the webview.
    let mut rest = use_signal(|| 0.0_f64);

    let start_script =
        use_gesture(
            SWIPE_SCRIPT,
            row_id.clone(),
            move |gesture| match SwipeEvent::parse(&gesture) {
                Some(SwipeEvent::Rest(offset)) => rest.set(offset),
                Some(SwipeEvent::Move(state)) => {
                    if let Some(on_swipe) = on_swipe {
                        on_swipe.call(state);
                    }
                }
                Some(SwipeEvent::Activate(state)) => {
                    if let Some(on_activate) = on_activate {
                        on_activate.call(state);
                    }
                }
                Some(SwipeEvent::Dismiss(state)) => {
                    if let Some(on_dismiss) = on_dismiss {
                        on_dismiss.call(state);
                    }
                }
                Some(SwipeEvent::LongPress) => {
                    if let Some(on_long_press) = on_long_press {
                        on_long_press.call(());
                    }
                }
                None => {}
            },
        );

    let current = rest();
    let open = current != 0.0;
    // Opens or closes the row: `"start"`, `"end"` or `"close"`.
    let command = use_callback({
        let row_id = row_id.clone();
        move |to: &'static str| gesture_command(SWIPE_SCRIPT, &row_id, to)
    });

    rsx! {
        div {
            id: row_id,
            class: merge_classes("g3-swipe-item", class.as_deref()),
            role: in_list.then_some("listitem"),
            onmounted: move |_| start_script.call(()),
            // Read by the script at each press, so a change takes effect on
            // the next gesture.
            "data-start-behavior": start_behavior.as_str(),
            "data-end-behavior": end_behavior.as_str(),
            "data-has-start": has_start.to_string(),
            "data-has-end": has_end.to_string(),
            "data-disabled": disabled.then_some("true"),
            // A mouse cannot drag this row, so its actions button shows on hover.
            "data-mouse-swipe": (!mouse_swipe).then_some("false"),
            "data-long-press": on_long_press.is_some().then_some("true"),
            "data-report-swipe": on_swipe.is_some().then_some("true"),
            onkeydown: move |event| {
                if disabled {
                    return;
                }
                match event.key() {
                    Key::ArrowLeft => command.call("end"),
                    Key::ArrowRight => command.call("start"),
                    Key::Escape if open => {
                        event.stop_propagation();
                        command.call("close");
                    }
                    _ => {}
                }
            },
            if let Some(actions) = start_actions {
                div {
                    class: "g3-swipe-actions g3-swipe-actions-start",
                    aria_hidden: (current <= 0.0).to_string(),
                    inert: (current <= 0.0).then_some(true),
                    // Pressing a revealed action finishes the swipe.
                    onclick: move |_| command.call("close"),
                    {actions}
                }
            }
            if let Some(actions) = end_actions {
                div {
                    class: "g3-swipe-actions g3-swipe-actions-end",
                    aria_hidden: (current >= 0.0).to_string(),
                    inert: (current >= 0.0).then_some(true),
                    onclick: move |_| command.call("close"),
                    {actions}
                }
            }
            div {
                class: "g3-swipe-content",
                {children}
                if (has_start || has_end) && !disabled {
                    button {
                        r#type: "button",
                        class: "g3-swipe-toggle",
                        aria_expanded: open.to_string(),
                        onclick: move |_| {
                            if open {
                                command.call("close");
                            } else if has_end {
                                command.call("end");
                            } else {
                                command.call("start");
                            }
                        },
                        "{strings.show_actions}"
                    }
                }
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn side_follows_direction() {
        assert_eq!(side_of(12.0), Some(SwipeSide::Start));
        assert_eq!(side_of(-12.0), Some(SwipeSide::End));
        assert_eq!(side_of(0.0), None);
    }

    #[test]
    fn reports_name_their_side() {
        let parse = |kind: &str, values: [f64; 3]| {
            SwipeEvent::parse(&Gesture {
                kind: kind.into(),
                values: values.into(),
            })
        };
        let Some(SwipeEvent::Activate(state)) = parse("activate", [-70.0, -0.5, 1.0]) else {
            panic!("an activation is parsed");
        };
        assert_eq!(state.side, SwipeSide::End);
        assert!(state.committed);
        assert_eq!(parse("rest", [0.0, 0.0, 0.0]), Some(SwipeEvent::Rest(0.0)));
        // Only a row that moved has a side to report.
        assert_eq!(parse("activate", [0.0, 0.0, 1.0]), None);
        assert_eq!(parse("unknown", [4.0, 0.0, 0.0]), None);
    }

    #[test]
    fn script_stays_alive_and_never_waits_on_rust() {
        super::super::gesture::assert_gesture_script(SWIPE_SCRIPT);
    }
}

File: src/components/swipe.js
// Drives every SwipeItem on the page from inside the webview.
//
// On native renderers each Rust event handler is a synchronous request from
// the webview to Rust, and the page cannot paint or scroll until it returns.
// Tracking a finger from Rust costs one of those per pointer move, which on
// Android is too slow for the row to follow the finger at all, and a Rust
// touchmove listener makes every scroll in the app wait on one. So the gesture
// runs here, and Rust hears only how each one ended. Each row describes itself
// in data attributes, which are read afresh at every press.
const key = Symbol.for("g3-ui.swipe");
window[key]?.dispose?.();

const REVEAL_WIDTH = 88;
const ACTIVATE_WIDTH = 136;
const DISMISS_WIDTH = 104;
const FULL_SWIPE_MARGIN = 30;
const ELASTIC_FACTOR = 0.55;
const ACTIVATE_RATIO = 0.48;
const ACTIVATE_SOFTENING = 0.72;
const DISMISS_OFFSET = 430;
const DISMISS_EXIT_MS = 560;
const DISMISS_COLLAPSE_MS = 180;
const LONG_PRESS_MS = 500;
const LONG_PRESS_SLOP = 8;
// Matches the browser's own slop: below it every gesture looks diagonal, and
// claiming one would steal ordinary scrolls.
const HORIZONTAL_SLOP = 10;
const SUPPRESS_CLICK_MS = 300;

// Where each row rests between gestures: 0, or open by an edge's width.
const rests = new WeakMap();
let active = null;
let suppress = null;

const send = (row, kind, offset, edge, committed) => {
    dioxus.send([row.id, kind, [offset, edge ? offset / edge.width : 0, committed ? 1 : 0]]);
};

// The edge a drag of `raw` pixels moves toward, if it has actions.
const edgeFor = (row, raw) => {
    const side = raw > 0 ? "start" : raw < 0 ? "end" : null;
    if (!side || row.dataset[side === "start" ? "hasStart" : "hasEnd"] !== "true") return null;
    const behavior = row.dataset[side === "start" ? "startBehavior" : "endBehavior"];
    let width = behavior === "activate" ? ACTIVATE_WIDTH : DISMISS_WIDTH;
    if (behavior === "reveal") {
        // Revealed actions are as wide as their buttons.
        const actions = row.querySelector(`:scope > .g3-swipe-actions-${side}`);
        width = Math.max(actions?.getBoundingClientRect().width || REVEAL_WIDTH, 1);
    }
    return { behavior, width };
};

// Where the row sits for a raw drag distance toward `edge`.
const offsetFor = (edge, raw) => {
    const limit = Math.max(edge.width, 1);
    if (edge.behavior === "reveal") return Math.min(Math.max(raw, -edge.width), edge.width);
    if (edge.behavior === "dismiss") {
        if (raw > limit) return limit + (raw - limit) * ELASTIC_FACTOR;
        if (raw < -limit) return -limit + (raw + limit) * ELASTIC_FACTOR;
        return raw;
    }
    const softenFrom = limit * ACTIVATE_SOFTENING;
    const distance = Math.abs(raw);
    if (distance <= softenFrom) return raw;
    const extra = distance - softenFrom;
    const remaining = Math.max(limit - softenFrom, 1);
    return Math.sign(raw) * (softenFrom + remaining * (extra / (extra + remaining)));
};

const committedAt = (edge, offset) => edge.behavior === "activate"
    ? Math.abs(offset) >= edge.width * ACTIVATE_RATIO
    : Math.abs(offset) >= edge.width + FULL_SWIPE_MARGIN;

const paint = (row, offset, dragging) => {
    const edge = edgeFor(row, offset);
    const width = edge?.width ?? REVEAL_WIDTH;
    row.style.setProperty("--g3-swipe-offset", `${offset}px`);
    row.style.setProperty("--g3-swipe-progress", String(Math.min(Math.abs(offset / width), 1.4)));
    // Which way a dismissed row leaves.
    row.style.setProperty("--g3-swipe-exit", offset < 0 ? "-1" : "1");
    if (dragging) row.dataset.dragging = "true";
    else delete row.dataset.dragging;
    // Rust hears only where a row comes to rest, so its `aria-hidden` still
    // says closed for the whole drag - and for all of an activate swipe, which
    // always rests shut. The edge being uncovered is shown from here instead.
    if (offset > 0) row.dataset.exposed = "start";
    else if (offset < 0) row.dataset.exposed = "end";
    else delete row.dataset.exposed;
    if (edge && edge.behavior !== "reveal" && committedAt(edge, offset)) row.dataset.committed = "true";
    else delete row.dataset.committed;
};

// Settle a row at rest and tell Rust, which keeps the hidden actions inert.
const rest = (row, offset) => {
    paint(row, offset, false);
    if ((rests.get(row) ?? 0) !== offset) {
        rests.set(row, offset);
        send(row, "rest", offset, edgeFor(row, offset), false);
    }
};

const idle = (row) => (row.dataset.state ?? "idle") === "idle";

// A drag that is clearly sideways must not scroll the page. Attached to the
// row itself, so only rows are blocking touch regions rather than the whole
// document, and the handler never leaves the webview.
const onTouchMove = (event) => {
    if (active?.horizontal && event.cancelable) event.preventDefault();
};

const onPointerDown = (event) => {
    // A reorder handle or a rating star in the row claimed this press first.
    if (event.defaultPrevented) return;
    const row = event.target instanceof Element && event.target.closest(".g3-swipe-item");
    if (!row || active || row.dataset.disabled === "true" || !idle(row)) return;
    if (row.dataset.mouseSwipe === "false" && event.pointerType === "mouse") return;
    if (!row.g3SwipeTouch) {
        row.g3SwipeTouch = true;
        row.addEventListener("touchmove", onTouchMove, { passive: false });
    }
    const base = rests.get(row) ?? 0;
    active = {
        row,
        pointerId: event.pointerId,
        x: event.clientX,
        y: event.clientY,
        base,
        offset: base,
        horizontal: false,
        moved: false,
        timer: null,
    };
    // Capture at once. Touch browsers can dispatch a leave as soon as the
    // finger moves off the original hit-test box; if capture waits for
    // horizontal intent, that leave ends the gesture and the row snaps back.
    try { row.setPointerCapture(event.pointerId); } catch (error) {}
    if (row.dataset.longPress === "true") {
        const gesture = active;
        gesture.timer = setTimeout(() => {
            if (active === gesture && !gesture.moved) send(row, "long-press", 0, null, false);
        }, LONG_PRESS_MS);
    }
};

const onPointerMove = (event) => {
    const gesture = active;
    if (!gesture || event.pointerId !== gesture.pointerId || !idle(gesture.row)) return;
    const moveX = event.clientX - gesture.x;
    const dy = event.clientY - gesture.y;
    if (Math.hypot(moveX, dy) > LONG_PRESS_SLOP) gesture.moved = true;
    if (!gesture.horizontal) {
        if (Math.abs(moveX) > HORIZONTAL_SLOP && Math.abs(moveX) > Math.abs(dy)) {
            gesture.horizontal = true;
        } else {
            return;
        }
    }
    // Resume from where an open row rests.
    const raw = gesture.base + moveX;
    const edge = edgeFor(gesture.row, raw);
    gesture.offset = edge ? offsetFor(edge, raw) : 0;
    paint(gesture.row, gesture.offset, true);
    if (edge && gesture.row.dataset.reportSwipe === "true") {
        send(gesture.row, "move", gesture.offset, edge, committedAt(edge, gesture.offset));
    }
};

const release = (event) => {
    const gesture = active;
    if (!gesture || event.pointerId !== gesture.pointerId) return;
    active = null;
    clearTimeout(gesture.timer);
    const { row, offset } = gesture;
    if (!gesture.horizontal) return;
    // The click that ends a swipe must not also activate the row.
    suppress = { row, until: performance.now() + SUPPRESS_CLICK_MS };
    const edge = edgeFor(row, offset);
    if (!edge) return rest(row, 0);
    const committed = committedAt(edge, offset);
    if (edge.behavior === "reveal") {
        return rest(row, Math.abs(offset) > edge.width / 2 ? Math.sign(offset) * edge.width : 0);
    }
    if (edge.behavior === "activate") {
        if (committed) send(row, "activate", offset, edge, true);
        return rest(row, 0);
    }
    if (!committed) return rest(row, 0);
    // Collapse from the row's own height, whatever it holds.
    row.style.setProperty("--g3-swipe-row-height", `${row.getBoundingClientRect().height}px`);
    paint(row, DISMISS_OFFSET * Math.sign(offset), false);
    row.dataset.state = "exiting";
    setTimeout(() => {
        row.dataset.state = "collapsing";
        setTimeout(() => send(row, "dismiss", offset, edge, true), DISMISS_COLLAPSE_MS);
    }, DISMISS_EXIT_MS);
};

const onPointerCancel = (event) => {
    const gesture = active;
    if (!gesture || event.pointerId !== gesture.pointerId) return;
    active = null;
    clearTimeout(gesture.timer);
    if (idle(gesture.row)) rest(gesture.row, 0);
};

const onClick = (event) => {
    if (!suppress || performance.now() > suppress.until) return;
    if (event.target instanceof Node && suppress.row.contains(event.target)) {
        event.preventDefault();
        event.stopPropagation();
    }
};

document.addEventListener("pointerdown", onPointerDown);
document.addEventListener("pointermove", onPointerMove);
document.addEventListener("pointerup", release);
document.addEventListener("lostpointercapture", release);
document.addEventListener("pointercancel", onPointerCancel);
// Capturing, so it runs before the renderer's own click listener.
document.addEventListener("click", onClick, true);

let released;
const done = new Promise((resolve) => {
    released = resolve;
});
window[key] = {
    // Opening or closing from the keyboard, the actions button, or a pressed
    // action.
    command(id, to) {
        const row = document.getElementById(id);
        if (!row || !idle(row)) return;
        if (to === "close") return rest(row, 0);
        // Whatever the edge's swipe does, this uncovers its buttons.
        if (row.dataset[to === "start" ? "hasStart" : "hasEnd"] !== "true") return;
        const actions = row.querySelector(`:scope > .g3-swipe-actions-${to}`);
        const width = Math.max(actions?.getBoundingClientRect().width || REVEAL_WIDTH, 1);
        rest(row, to === "start" ? width : -width);
    },
    dispose() {
        document.removeEventListener("pointerdown", onPointerDown);
        document.removeEventListener("pointermove", onPointerMove);
        document.removeEventListener("pointerup", release);
        document.removeEventListener("lostpointercapture", release);
        document.removeEventListener("pointercancel", onPointerCancel);
        document.removeEventListener("click", onClick, true);
        released();
    },
};
// Native renderers close an eval's channel as soon as its script returns.
await done;
